### PR TITLE
[FLINK-35993][table] Add the built-in function UNHEX

### PR DESCRIPTION
**Pull request by dylanhz**
## What is the purpose of the change

Add the built-in function UNHEX.
Examples:
```SQL
> SELECT DECODE(UNHEX('466C696E6B'), 'UTF-8');
Flink
```

## Brief change log

[FLINK-35993](https://issues.apache.org/jira/browse/FLINK-35993)

## Verifying this change

`MathFunctionsITCase#unhexTestCases`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/MathFunctionsITCase.java
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/MathFunctionsITCase.java:182]: Add a case: UNHEX(HEX('😀')

File: docs/data/sql_functions.yml
Comment (lincoln-lil) [inline docs/data/sql_functions.yml:246]: Add a blank line before and after this one to make it look a little clearer.

**Review by lincoln-lil — COMMENTED** [review]
@dylanhz Overall looks good to me! I left some comments, PTAL.
Btw, I found a doc issue of previously added function TRANSLATE:
image
https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/functions/systemfunctions

We can include the fix within this pr.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java
Comment (lincoln-lil) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:1868]: Should use forceNullable here.

Comment (dylanhz) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:1868]: DataType created by these methods is nullable by default.

Comment (lincoln-lil) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:1919]: You're right : )

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/MathFunctionsITCase.java
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/MathFunctionsITCase.java:220]: Cool! Verified this can get the expected 😀

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java
Comment (lincoln-lil) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java:191]: It would be good to emphasize specific processing differences, including header byte strategies for odd-length inputs.
Another thing is the method name, I'm thinking about using a consistent name, e.g., `decodeHexLax` or `lenientHexDecode `, WDYT?

Comment (lincoln-lil) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java:191]: Perhaps we can phrase like this:
```
 * Converts an array of characters representing hexadecimal values into an array of bytes of
 * those same values. The returned array will be half the length of the passed array, as it
 * takes two characters to represent any given byte. If the input array has an odd length,
 * the first byte is handled separately and set to 0.
 *
 * <p>Unlike {@link #decodeHex(String)}, this method does not throw an exception for odd-length
 * inputs or invalid characters. Instead, it returns null if invalid characters are encountered.
 *
 * @param bytes An array of characters containing hexadecimal digits.
 * @return A byte array containing the binary data decoded from the supplied char array,
 *         or null if the input contains invalid hexadecimal characters.
```

**Review by lincoln-lil — COMMENTED** [review]
@dylanhz thanks for your updates! Just two comments before we merge it.
And you can squash the commits into two, one for the new unhex, and another for the doc's fix.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java
Comment (dylanhz) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java:191]: Should we keep consistent with existed method `hex`, as with the built-in functions `HEX` and `UNHEX`?
And maybe we can create a function `unhex(byte[] bytes, boolean throwOnFailure)` to call both implementations and change the signature of `decodeHex` to `byte[]`, because in the original implementation `String` is also converted to a char array for calling. This also ensures performance for `BinaryStringData`.

**Review by lincoln-lil — APPROVED** [review]
@dylanhz Thanks for the updates! LGTM!
Waiting for ci green before merging.

**Pull request merged**
